File: backend/pkg/model/game.go
package model

import "time"

type Game struct {
	Id        int64      `db:"id" json:"id"`
	UserId    int64      `db:"user_id" json:"user_id"`
	WordId    int64      `db:"word_id" json:"word_id"`
	Status    string     `db:"status" json:"status"`
	Language  string     `db:"language" json:"language"`
	StartedAt time.Time  `db:"started_at" json:"started_at"`
	EndedAt   *time.Time `db:"ended_at" json:"ended_at,omitempty"`
	ExpiresAt *time.Time `db:"expires_at" json:"expires_at,omitempty"`
}

type Guess struct {
	Id        int64     `db:"id" json:"id"`
	GameId    int64     `db:"game_id" json:"game_id"`
	GuessWord string    `db:"guess_word" json:"guess_word"`
	Distance  int       `db:"distance" json:"distance"`
	CreatedAt time.Time `db:"created_at" json:"created_at"`
}

type GuessResult struct {
	Rank         int  `json:"rank"`
	Found        bool `json:"found"`
	InVocabulary bool `json:"in_vocabulary"`
}

File: frontend/src/types/game.ts
import type { ComponentType } from 'react';

export type GameStatus = 'in_progress' | 'won' | 'lost' | 'abandoned';

export type GameModeId = 'daily' | 'competitive' | 'endless' | 'time-attack' | 'practice';

export type MultiplayerModeId = '1v1-duel' | 'arena';

export interface Game {
    id: number;
    language: string;
    status: GameStatus;
    started_at: string;
    finished_at?: string;
    mode?: GameModeId;
}

export interface Guess {
    id: number;
    game_id: number;
    guess_word: string;
    distance: number;
    created_at: string;
}

export interface GuessResult {
    rank: number;
    found: boolean;
    in_vocabulary: boolean;
}

export interface GameState {
    gameId: number;
    guesses: Guess[];
    language: string;
    status: GameStatus;
    mode: GameModeId;
    savedAt?: string;
}

export interface GameMode {
    id: GameModeId;
    name: string;
    description: string;
    icon: ComponentType<{ className?: string }>;
    available: boolean;
    path: string;
    gradient: string;
    rated: boolean;
    persistent: boolean;
    canContinue: boolean;
    concurrent: boolean;
    hasLeaderboard: boolean;
    comingSoon?: boolean;
}

export interface MultiplayerMode {
    id: MultiplayerModeId;
    name: string;
    description: string;
    icon: ComponentType<{ className?: string }>;
    comingSoon: boolean;
}
